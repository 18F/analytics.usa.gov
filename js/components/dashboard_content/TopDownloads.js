import React, { useRef, useEffect, useState } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import ChartBuilder from "../../lib/chart_helpers/chart_builder";
import DataLoader from "../../lib/data_loader";
import barChart from "../../lib/chart_helpers/barchart";
import formatters from "../../lib/chart_helpers/formatters";
import FilterSelect from "../select/FilterSelect";

/**
 * Retrieves the top downloads report from the passed data URL and creates a
 * visualization for the count of users downloading files for the current
 * agency.
 *
 * @param {object} props the properties for the component
 * @param {string} props.agency the agency display name to be used in the chart
 * description.
 * @param {string} props.dataHrefBase the URL of the base location of the data
 * to be downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @param {number} props.numberOfListingsToDisplay the count of downloads
 * listings to display in the bar chart.
 * @returns {import('react').ReactElement} The rendered element
 */
function TopDownloads({ agency, dataHrefBase, numberOfListingsToDisplay }) {
  const reportFilters = [
    ["Yesterday", "top-downloads-yesterday"],
    ["7 Days", "top-downloads-7-days"],
    ["30 Days", "top-downloads-30-days"],
    ["90 Days", "top-downloads-90-days"],
  ];
  const [currentFilter, setCurrentFilter] = useState(reportFilters[0]);
  const [shouldDisplayDownloads, setShouldDisplayDownloads] = useState(true);
  const ref = useRef(null);

  useEffect(() => {
    const initChart = async () => {
      if (currentFilter) {
        await loadDataAndBuildChart();
      }
    };
    initChart().catch(console.error);
  }, [currentFilter]);

  async function loadDataAndBuildChart() {
    let data;

    try {
      data = await DataLoader.loadDailyReportJSON(
        `${dataHrefBase}/${currentFilter[1]}.json`,
      );
    } catch (e) {
      data = { totals: {} };
    }
    await buildChartForData(data);
  }

  async function buildChartForData(data) {
    let downloadData;
    if (data && data.data && Array.isArray(data.data) && data.data.length > 2) {
      downloadData = data;
      setShouldDisplayDownloads(true);
    } else {
      downloadData = { data: [] };
      setShouldDisplayDownloads(false);
    }

    const chartBuilder = new ChartBuilder();
    await chartBuilder
      .setElement(ref.current)
      .setData(downloadData)
      .setTransformer((d) => d.data.slice(0, numberOfListingsToDisplay))
      .setRenderer(
        barChart()
          .value((d) => +d.total_events)
          .label(
            (d) =>
              `<div class="grid-row">
                <div class="grid-col-12 text--overflow-ellipsis">
                  <a target="_blank" rel="noopener" href="http://${d.page}">
                    ${d.page_title}
                  </a>
                </div>
                <div class="grid-col-12">
                  <div class="grid-row text--overflow-wrap">
                    <div class="grid-col-auto text--lowercase dark-gray">
                      <span>${formatters.formatURL(d.page).trim()}</span><span class="padding-x-05">/</span>
                    </div>
                    <div class="grid-col-auto text--lowercase">
                      <a target="_blank" rel="noopener" href="${d.linkUrl}">
                        download file
                      </a>
                    </div>
                  </div>
                </div>
              </div>`,
          )
          .scale((values) =>
            d3.scale
              .linear()
              .domain([0, 1, d3.max(values)])
              .rangeRound([0, 1, 100]),
          )
          .format(formatters.addCommas),
      )
      .build();
  }

  async function filterChangeHandler(fileName) {
    if (!fileName) return;

    const selectedFilter = reportFilters.find((reportFilter) => {
      return reportFilter[1] == fileName;
    });
    await setCurrentFilter(selectedFilter);
  }

  function timeIntervalDescription() {
    if (currentFilter[0] == "Yesterday") {
      return currentFilter[0].toLowerCase();
    } else {
      return `over the last ${currentFilter[0].toLowerCase()}`;
    }
  }

  return (
    <>
      <p className="margin-top-0 margin-bottom-1">
        <em>
          {shouldDisplayDownloads
            ? `Top files downloaded ${timeIntervalDescription()} on ${agency} hostnames.`
            : `Top downloads data ${timeIntervalDescription()} is unavailable for ${agency} hostnames.`}
        </em>
      </p>
      <div className="grid-row">
        <div className="display-flex card:grid-col-12 mobile-lg:grid-col-fill card:flex-justify-center mobile-lg:flex-justify-start card:padding-bottom-105 mobile-lg:padding-bottom-0">
          {shouldDisplayDownloads && (
            <p className="margin-top-05 margin-bottom-05">
              <a
                href={`${dataHrefBase}/${currentFilter[1]}.csv`}
                aria-label={`${currentFilter[1]}.csv`}
              >
                Download the data
                <svg
                  className="usa-icon margin-bottom-neg-05 margin-left-05"
                  aria-hidden="true"
                  focusable="false"
                  role="img"
                >
                  <use xlinkHref="/assets/uswds/img/sprite.svg#file_present"></use>
                </svg>
              </a>
            </p>
          )}
        </div>
        <div className="card:grid-col-12 mobile-lg:grid-col-auto">
          <div className="display-flex card:flex-justify-center mobile-lg:flex-justify-end">
            <FilterSelect
              filters={reportFilters}
              defaultFilterValue={reportFilters[0][1] || ""}
              onChange={filterChangeHandler}
              name={"top downloads chart time filter"}
            />
          </div>
        </div>
      </div>
      {shouldDisplayDownloads && (
        <figure className="top-downloads__bar-chart" ref={ref}>
          <div className="data chart__bar-chart text--capitalize margin-top-2"></div>
        </figure>
      )}
    </>
  );
}

TopDownloads.propTypes = {
  agency: PropTypes.string.isRequired,
  dataHrefBase: PropTypes.string.isRequired,
  numberOfListingsToDisplay: PropTypes.number.isRequired,
};

export default TopDownloads;
