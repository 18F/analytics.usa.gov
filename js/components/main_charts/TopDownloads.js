import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import barChart from "../../lib/chart_helpers/barchart";
import formatters from "../../lib/chart_helpers/formatters";
import renderBlock from "../../lib/chart_helpers/renderblock";

/**
 * Retrieves the top downloads report from the passed data URL and creates a
 * visualization for the count of users downloading files for the current
 * agency.
 *
 * @param {String} dataHrefBase the URL of the base location of the data to be
 * downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 */
function TopDownloads({ dataHrefBase }) {
  const reportURL = `${dataHrefBase}/top-downloads-yesterday.json`;
  const ref = useRef(null);

  useEffect(() => {
    const initTopDownloads = async () => {
      const result = await d3
        .select(ref.current)
        .datum({
          source: reportURL,
          block: ref.current,
        })
        .call(
          renderBlock
            .loadAndRender()
            .transform((d) => d.data.slice(0, 10))
            .render(
              barChart()
                .value((d) => +d.total_events)
                .label((d) =>
                  [
                    '<span class="name"><a class="top-download-page" target="_blank" rel="noopener" href=http://',
                    d.page,
                    ">",
                    d.page_title,
                    "</a></span> ",
                    '<span class="domain" >',
                    formatters.formatURL(d.page),
                    "</span> ",
                    '<span class="divider">/</span> ',
                    '<span class="filename"><a class="top-download-file" target="_blank" aria-label="',
                    formatters.formatFile(d.file_name),
                    '" rel="noopener" href=',
                    formatters.formatProtocol(d.page),
                    formatters.formatFile(d.file_name),
                    ">",
                    "download file",
                    "</a></span>",
                  ].join(""),
                )
                .scale((values) =>
                  d3.scale
                    .linear()
                    .domain([0, 1, d3.max(values)])
                    .rangeRound([0, 1, 100]),
                )
                .format(formatters.addCommas),
            ),
        );
      return result;
    };
    initTopDownloads().catch(console.error);
  }, []);

  return (
    <figure id="top-downloads" ref={ref}>
      <div className="data bar-chart"></div>
    </figure>
  );
}

TopDownloads.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default TopDownloads;
