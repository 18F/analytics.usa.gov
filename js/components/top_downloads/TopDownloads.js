import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import barChart from "../../lib/chart_helpers/barchart";
import formatters from "../../lib/chart_helpers/formatters";
import renderBlock from "../../lib/chart_helpers/renderblock";

function TopDownloads({ dataURL, dataPrefix }) {
  const reportURL = `${dataURL}/${dataPrefix}/top-downloads-yesterday.json`;
  const ref = useRef(null);

  useEffect(() => {
    const initTopDownloads = async () => {
      const result = await d3
        .select(ref.current)
        .datum({
          source: reportURL,
          block: ref.current,
        })
        .call(
          renderBlock
            .loadAndRender()
            .transform((d) => d.data.slice(0, 10))
            .render(
              barChart()
                .value((d) => +d.total_events)
                .label((d) =>
                  [
                    '<span class="name"><a class="top-download-page" target="_blank" rel="noopener" href=http://',
                    d.page,
                    ">",
                    d.page_title,
                    "</a></span> ",
                    '<span class="domain" >',
                    formatters.formatURL(d.page),
                    "</span> ",
                    '<span class="divider">/</span> ',
                    '<span class="filename"><a class="top-download-file" target="_blank" aria-label="',
                    formatters.formatFile(d.file_name),
                    '" rel="noopener" href=',
                    formatters.formatProtocol(d.page),
                    formatters.formatFile(d.file_name),
                    ">",
                    "download file",
                    "</a></span>",
                  ].join(""),
                )
                .scale((values) =>
                  d3.scale
                    .linear()
                    .domain([0, 1, d3.max(values)])
                    .rangeRound([0, 1, 100]),
                )
                .format(formatters.addCommas),
            ),
        );
      return result;
    };
    initTopDownloads().catch(console.error);
  }, []);

  return (
    <figure id="top-downloads" ref={ref}>
      <div className="data bar-chart"></div>
    </figure>
  );
}

TopDownloads.propTypes = {
  dataURL: PropTypes.string.isRequired,
  dataPrefix: PropTypes.string.isRequired,
};

export default TopDownloads;
