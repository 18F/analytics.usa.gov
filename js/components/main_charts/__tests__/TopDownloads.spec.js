import React from "react";
import { render, screen, waitFor } from "@testing-library/react";

import DataLoader from "../../../lib/data_loader";
import TopDownloads from "../TopDownloads";

jest.mock("../../../lib/data_loader", () => ({
  ...jest.requireActual("../../../lib/data_loader"),
  loadJSON: jest.fn(),
}));

describe("TopDownloads", () => {
  let component;
  let data;

  describe("when data is not loaded", () => {
    beforeEach(() => {
      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.resolve(null);
      });
      component = render(
        <TopDownloads
          dataHrefBase="http://www.example.com/data/"
          reportFileName="foobar.json"
          numberOfListingsToDisplay={20}
        />,
      );
    });

    it("renders a component in loading state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data is loaded", () => {
    beforeEach(async () => {
      data = {
        name: "top-downloads-yesterday",
        query: {},
        meta: {
          name: "Top Downloads Yesterday",
          description: "Top downloads yesterday",
        },
        data: [
          {
            date: "2024-01-04",
            page_title:
              "Inactivated Influenza Vaccine Information Statement | CDC",
            event_label: "file_download",
            file_name:
              "https://www.cdc.gov/vaccines/hcp/vis/vis-statements/flu.pdf",
            page: "www.cdc.gov/vaccines/hcp/vis/vis-statements/flu.html",
            total_events: "30000000",
          },
          {
            date: "2024-01-04",
            page_title:
              "About Form W-9, Request for Taxpayer Identification Number and Certification | Internal Revenue Service",
            event_label: "file_download",
            file_name: "/pub/irs-pdf/fw9.pdf",
            page: "www.irs.gov/forms-pubs/about-form-w-9",
            total_events: "28634",
          },
          {
            date: "2024-01-04",
            page_title: "Small Entity Compliance Guide | FinCEN.gov",
            event_label: "file_download",
            file_name:
              "/sites/default/files/shared/boi_small_compliance_guide.v1.1-final.pdf",
            page: "www.fincen.gov/boi/small-entity-compliance-guide",
            total_events: "9426",
          },
          {
            date: "2024-01-04",
            page_title: "Passport Fees",
            event_label: "file_download",
            file_name:
              "/content/dam/passports/forms-fees/fee%20chart%202023.pdf",
            page: "travel.state.gov/content/travel/en/passports/how-apply/fees.html",
            total_events: "9295",
          },
          {
            date: "2024-01-04",
            page_title: "Employment Eligibility Verification | USCIS",
            event_label: "file_download",
            file_name: "/sites/default/files/document/forms/i-9.pdf",
            page: "www.uscis.gov/i-9",
            total_events: "8984",
          },
          {
            date: "2024-01-04",
            page_title: "General Schedule",
            event_label: "file_download",
            file_name:
              "/policy-data-oversight/pay-leave/salaries-wages/salary-tables/24tables/pdf/dcb.pdf",
            page: "www.opm.gov/policy-data-oversight/pay-leave/salaries-wages/2024/general-schedule",
            total_events: "8774",
          },
          {
            date: "2024-01-04",
            page_title: "General Schedule",
            event_label: "file_download",
            file_name:
              "/policy-data-oversight/pay-leave/salaries-wages/salary-tables/24tables/pdf/rus.pdf",
            page: "www.opm.gov/policy-data-oversight/pay-leave/salaries-wages/2024/general-schedule",
            total_events: "8174",
          },
          {
            date: "2024-01-04",
            page_title:
              "About Form W-4, Employee's Withholding Certificate | Internal Revenue Service",
            event_label: "file_download",
            file_name: "/pub/irs-pdf/fw4.pdf",
            page: "www.irs.gov/forms-pubs/about-form-w-4",
            total_events: "6385",
          },
        ],
        totals: {},
        taken_at: "2024-01-05T15:31:55.633Z",
      };

      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.resolve(data);
      });
      component = render(
        <TopDownloads
          dataHrefBase="http://www.example.com/data/"
          reportFileName="foobar.json"
          numberOfListingsToDisplay={20}
        />,
      );
      await waitFor(() => screen.getByText("Passport Fees"));
    });

    it("renders a component with data loaded", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data loading has an error", () => {
    const error = "you broke it";

    beforeEach(() => {
      console.error = jest.fn();
      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.reject(error);
      });
      component = render(
        <TopDownloads
          dataHrefBase="http://www.example.com/data/"
          reportFileName="foobar.json"
          numberOfListingsToDisplay={20}
        />,
      );
    });

    it("renders a component in error state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });

    it("logs the error to console", () => {
      expect(console.error).toHaveBeenCalledWith(error);
    });
  });
});
