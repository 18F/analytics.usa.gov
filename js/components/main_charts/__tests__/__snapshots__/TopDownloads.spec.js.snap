// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopDownloads when data is loaded renders a component with data loaded 1`] = `
<DocumentFragment>
  <figure
    class="top-downloads__bar-chart"
  >
    <div
      class="data bar-chart"
    >
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <span
            class="name"
          >
            <a
              class="top-download-page"
              href="http://www.cdc.gov/vaccines/hcp/vis/vis-statements/flu.html"
              rel="noopener"
              target="_blank"
            >
              Inactivated Influenza Vaccine Information Statement | CDC
            </a>
          </span>
           
          <span
            class="domain"
          >
            www.cdc.gov
          </span>
           
          <span
            class="divider"
          >
            /
          </span>
           
          <span
            class="filename"
          >
            <a
              aria-label="/vaccines/hcp/vis/vis-statements/flu.pdf"
              class="top-download-file"
              href="https://www.cdc.gov/vaccines/hcp/vis/vis-statements/flu.pdf"
              rel="noopener"
              target="_blank"
            >
              download file
            </a>
          </span>
        </div>
        <div
          class="value"
        >
          30,000,000
        </div>
        <div
          class="bar"
          style="width: 100.0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <span
            class="name"
          >
            <a
              class="top-download-page"
              href="http://www.irs.gov/forms-pubs/about-form-w-9"
              rel="noopener"
              target="_blank"
            >
              About Form W-9, Request for Taxpayer Identification Number and Certification | Internal Revenue Service
            </a>
          </span>
           
          <span
            class="domain"
          >
            www.irs.gov
          </span>
           
          <span
            class="divider"
          >
            /
          </span>
           
          <span
            class="filename"
          >
            <a
              aria-label="/pub/irs-pdf/fw9.pdf"
              class="top-download-file"
              href="https://www.irs.gov/pub/irs-pdf/fw9.pdf"
              rel="noopener"
              target="_blank"
            >
              download file
            </a>
          </span>
        </div>
        <div
          class="value"
        >
          28,634
        </div>
        <div
          class="bar"
          style="width: 1.0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <span
            class="name"
          >
            <a
              class="top-download-page"
              href="http://www.fincen.gov/boi/small-entity-compliance-guide"
              rel="noopener"
              target="_blank"
            >
              Small Entity Compliance Guide | FinCEN.gov
            </a>
          </span>
           
          <span
            class="domain"
          >
            www.fincen.gov
          </span>
           
          <span
            class="divider"
          >
            /
          </span>
           
          <span
            class="filename"
          >
            <a
              aria-label="/sites/default/files/shared/boi_small_compliance_guide.v1.1-final.pdf"
              class="top-download-file"
              href="https://www.fincen.gov/sites/default/files/shared/boi_small_compliance_guide.v1.1-final.pdf"
              rel="noopener"
              target="_blank"
            >
              download file
            </a>
          </span>
        </div>
        <div
          class="value"
        >
          9,426
        </div>
        <div
          class="bar"
          style="width: 1.0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <span
            class="name"
          >
            <a
              class="top-download-page"
              href="http://travel.state.gov/content/travel/en/passports/how-apply/fees.html"
              rel="noopener"
              target="_blank"
            >
              Passport Fees
            </a>
          </span>
           
          <span
            class="domain"
          >
            travel.state.gov
          </span>
           
          <span
            class="divider"
          >
            /
          </span>
           
          <span
            class="filename"
          >
            <a
              aria-label="/content/dam/passports/forms-fees/fee%20chart%202023.pdf"
              class="top-download-file"
              href="https://travel.state.gov/content/dam/passports/forms-fees/fee%20chart%202023.pdf"
              rel="noopener"
              target="_blank"
            >
              download file
            </a>
          </span>
        </div>
        <div
          class="value"
        >
          9,295
        </div>
        <div
          class="bar"
          style="width: 1.0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <span
            class="name"
          >
            <a
              class="top-download-page"
              href="http://www.uscis.gov/i-9"
              rel="noopener"
              target="_blank"
            >
              Employment Eligibility Verification | USCIS
            </a>
          </span>
           
          <span
            class="domain"
          >
            www.uscis.gov
          </span>
           
          <span
            class="divider"
          >
            /
          </span>
           
          <span
            class="filename"
          >
            <a
              aria-label="/sites/default/files/document/forms/i-9.pdf"
              class="top-download-file"
              href="https://www.uscis.gov/sites/default/files/document/forms/i-9.pdf"
              rel="noopener"
              target="_blank"
            >
              download file
            </a>
          </span>
        </div>
        <div
          class="value"
        >
          8,984
        </div>
        <div
          class="bar"
          style="width: 1.0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <span
            class="name"
          >
            <a
              class="top-download-page"
              href="http://www.opm.gov/policy-data-oversight/pay-leave/salaries-wages/2024/general-schedule"
              rel="noopener"
              target="_blank"
            >
              General Schedule
            </a>
          </span>
           
          <span
            class="domain"
          >
            www.opm.gov
          </span>
           
          <span
            class="divider"
          >
            /
          </span>
           
          <span
            class="filename"
          >
            <a
              aria-label="/policy-data-oversight/pay-leave/salaries-wages/salary-tables/24tables/pdf/dcb.pdf"
              class="top-download-file"
              href="https://www.opm.gov/policy-data-oversight/pay-leave/salaries-wages/salary-tables/24tables/pdf/dcb.pdf"
              rel="noopener"
              target="_blank"
            >
              download file
            </a>
          </span>
        </div>
        <div
          class="value"
        >
          8,774
        </div>
        <div
          class="bar"
          style="width: 1.0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <span
            class="name"
          >
            <a
              class="top-download-page"
              href="http://www.opm.gov/policy-data-oversight/pay-leave/salaries-wages/2024/general-schedule"
              rel="noopener"
              target="_blank"
            >
              General Schedule
            </a>
          </span>
           
          <span
            class="domain"
          >
            www.opm.gov
          </span>
           
          <span
            class="divider"
          >
            /
          </span>
           
          <span
            class="filename"
          >
            <a
              aria-label="/policy-data-oversight/pay-leave/salaries-wages/salary-tables/24tables/pdf/rus.pdf"
              class="top-download-file"
              href="https://www.opm.gov/policy-data-oversight/pay-leave/salaries-wages/salary-tables/24tables/pdf/rus.pdf"
              rel="noopener"
              target="_blank"
            >
              download file
            </a>
          </span>
        </div>
        <div
          class="value"
        >
          8,174
        </div>
        <div
          class="bar"
          style="width: 1.0%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          <span
            class="name"
          >
            <a
              class="top-download-page"
              href="http://www.irs.gov/forms-pubs/about-form-w-4"
              rel="noopener"
              target="_blank"
            >
              About Form W-4, Employee's Withholding Certificate | Internal Revenue Service
            </a>
          </span>
           
          <span
            class="domain"
          >
            www.irs.gov
          </span>
           
          <span
            class="divider"
          >
            /
          </span>
           
          <span
            class="filename"
          >
            <a
              aria-label="/pub/irs-pdf/fw4.pdf"
              class="top-download-file"
              href="https://www.irs.gov/pub/irs-pdf/fw4.pdf"
              rel="noopener"
              target="_blank"
            >
              download file
            </a>
          </span>
        </div>
        <div
          class="value"
        >
          6,385
        </div>
        <div
          class="bar"
          style="width: 1.0%;"
        />
      </div>
    </div>
  </figure>
</DocumentFragment>
`;

exports[`TopDownloads when data is not loaded renders a component in loading state 1`] = `
<DocumentFragment>
  <figure
    class="top-downloads__bar-chart"
  >
    <div
      class="data bar-chart"
    />
  </figure>
</DocumentFragment>
`;

exports[`TopDownloads when data loading has an error renders a component in error state 1`] = `
<DocumentFragment>
  <figure
    class="top-downloads__bar-chart"
  >
    <div
      class="data bar-chart"
    />
  </figure>
</DocumentFragment>
`;
